File: cli/internal/sdd/statemachine.go
package sdd

// stateMachine is the computed phase state for a concrete change.
type stateMachine struct {
	applyState   ApplyState
	dependencies Dependencies
	next         string
	reasons      []string
}

// computeStateMachine derives the apply state, per-phase dependencies, the next
// recommended action, and the blocked reasons from the artifact states, task
// progress, and whether the verify report is clearly passing.
func computeStateMachine(artifacts map[string]ArtifactState, tasks TaskProgress, verifyPassing bool) stateMachine {
	coreReady := isCoreReady(artifacts, tasks)
	applyState := resolveApplyState(coreReady, tasks)

	reasons := coreBlockedReasons(artifacts, tasks)
	if artifacts["verifyReport"] == ArtifactDone && !verifyPassing && applyState != ApplyReady {
		reasons = append(reasons, "verify-report.md is not clearly passing.")
	}

	dependencies := resolveDependencies(artifacts, tasks, applyState, coreReady, verifyPassing)
	next := resolveNextRecommended(dependencies, applyState)

	return stateMachine{
		applyState:   applyState,
		dependencies: dependencies,
		next:         next,
		reasons:      reasons,
	}
}

// isCoreReady reports whether the four core artifacts are done and tasks.md has
// at least one checkbox — the precondition for the apply phase.
func isCoreReady(artifacts map[string]ArtifactState, tasks TaskProgress) bool {
	return artifacts["proposal"] == ArtifactDone &&
		artifacts["specs"] == ArtifactDone &&
		artifacts["design"] == ArtifactDone &&
		artifacts["tasks"] == ArtifactDone &&
		tasks.Total > 0
}

// resolveApplyState classifies the apply phase: blocked until the core is
// ready, all_done when every task is checked, otherwise ready.
func resolveApplyState(coreReady bool, tasks TaskProgress) ApplyState {
	if !coreReady {
		return ApplyBlocked
	}
	if tasks.AllComplete {
		return ApplyAllDone
	}
	return ApplyReady
}

// resolveDependencies maps every phase to its readiness. The four core
// artifacts map done->all_done else blocked. Apply mirrors applyState. Verify
// and archive follow the cross-phase gating rules.
func resolveDependencies(artifacts map[string]ArtifactState, tasks TaskProgress, applyState ApplyState, coreReady, verifyPassing bool) Dependencies {
	deps := Dependencies{
		Proposal: artifactDependency(artifacts["proposal"]),
		Specs:    artifactDependency(artifacts["specs"]),
		Design:   artifactDependency(artifacts["design"]),
		Tasks:    artifactDependency(artifacts["tasks"]),
		Apply:    applyDependency(applyState),
		Verify:   DependencyBlocked,
		Archive:  DependencyBlocked,
	}

	applyProgressDone := artifacts["applyProgress"] == ArtifactDone
	verifyReportDone := artifacts["verifyReport"] == ArtifactDone

	if verifyReportDone && coreReady && tasks.AllComplete && verifyPassing {
		deps.Verify = DependencyAllDone
	} else if coreReady && (applyState == ApplyAllDone || applyProgressDone) {
		deps.Verify = DependencyReady
	}

	if deps.Verify == DependencyAllDone && tasks.AllComplete {
		deps.Archive = DependencyReady
	}
	return deps
}

// artifactDependency maps a core artifact state to its phase dependency.
func artifactDependency(state ArtifactState) DependencyState {
	if state == ArtifactDone {
		return DependencyAllDone
	}
	return DependencyBlocked
}

// applyDependency mirrors the internal apply state onto its phase dependency.
func applyDependency(applyState ApplyState) DependencyState {
	switch applyState {
	case ApplyReady:
		return DependencyReady
	case ApplyAllDone:
		return DependencyAllDone
	default:
		return DependencyBlocked
	}
}

// resolveNextRecommended chooses the single next action: apply when ready, else
// verify when ready, else archive when verify is done and apply is done, else
// resolve-blockers.
func resolveNextRecommended(deps Dependencies, applyState ApplyState) string {
	switch {
	case deps.Apply == DependencyReady:
		return phaseApply
	case deps.Verify == DependencyReady:
		return phaseVerify
	case deps.Verify == DependencyAllDone && applyState == ApplyAllDone:
		return phaseArchive
	default:
		return nextResolveBlockers
	}
}

// coreBlockedReasons lists the human reasons the core is not ready: one per
// missing-or-partial core artifact, plus a note when tasks.md is present but
// has no checkboxes.
func coreBlockedReasons(artifacts map[string]ArtifactState, tasks TaskProgress) []string {
	var reasons []string
	if artifacts["proposal"] != ArtifactDone {
		reasons = append(reasons, "proposal.md is missing or partial.")
	}
	if artifacts["specs"] != ArtifactDone {
		reasons = append(reasons, "specs/**/spec.md is missing or partial.")
	}
	if artifacts["design"] != ArtifactDone {
		reasons = append(reasons, "design.md is missing or partial.")
	}
	if artifacts["tasks"] != ArtifactDone {
		reasons = append(reasons, "tasks.md is missing or partial.")
	}
	if artifacts["tasks"] == ArtifactDone && tasks.Total == 0 {
		reasons = append(reasons, "tasks.md has no markdown task checkboxes.")
	}
	return reasons
}

File: cli/src/ai_harness/sdd/statemachine.py
"""Phase state machine: dependencies, apply state, next action, blockers."""

from __future__ import annotations

from dataclasses import dataclass

from .models import (
    APPLY_ALL_DONE,
    APPLY_BLOCKED,
    APPLY_READY,
    ARTIFACT_DONE,
    DEP_ALL_DONE,
    DEP_BLOCKED,
    DEP_READY,
    NEXT_RESOLVE_BLOCKERS,
    PHASE_APPLY,
    PHASE_ARCHIVE,
    PHASE_VERIFY,
    Dependencies,
    TaskProgress,
)


@dataclass
class StateMachine:
    apply_state: str
    dependencies: Dependencies
    next: str
    reasons: list[str]


def compute_state_machine(
    artifacts: dict[str, str], tasks: TaskProgress, verify_passing: bool
) -> StateMachine:
    """Derive the apply state, per-phase dependencies, the next recommended
    action, and the blocked reasons from the artifact states, task progress, and
    whether the verify report is clearly passing."""
    core_ready = _is_core_ready(artifacts, tasks)
    apply_state = _resolve_apply_state(core_ready, tasks)

    reasons = _core_blocked_reasons(artifacts, tasks)
    if (
        artifacts["verifyReport"] == ARTIFACT_DONE
        and not verify_passing
        and apply_state != APPLY_READY
    ):
        reasons.append("verify-report.md is not clearly passing.")

    dependencies = _resolve_dependencies(artifacts, tasks, apply_state, core_ready, verify_passing)
    next_recommended = _resolve_next_recommended(dependencies, apply_state)

    return StateMachine(
        apply_state=apply_state,
        dependencies=dependencies,
        next=next_recommended,
        reasons=reasons,
    )


def _is_core_ready(artifacts: dict[str, str], tasks: TaskProgress) -> bool:
    """The four core artifacts are done and tasks.md has at least one checkbox."""
    return (
        artifacts["proposal"] == ARTIFACT_DONE
        and artifacts["specs"] == ARTIFACT_DONE
        and artifacts["design"] == ARTIFACT_DONE
        and artifacts["tasks"] == ARTIFACT_DONE
        and tasks.total > 0
    )


def _resolve_apply_state(core_ready: bool, tasks: TaskProgress) -> str:
    if not core_ready:
        return APPLY_BLOCKED
    if tasks.all_complete:
        return APPLY_ALL_DONE
    return APPLY_READY


def _resolve_dependencies(
    artifacts: dict[str, str],
    tasks: TaskProgress,
    apply_state: str,
    core_ready: bool,
    verify_passing: bool,
) -> Dependencies:
    deps = Dependencies(
        proposal=_artifact_dependency(artifacts["proposal"]),
        specs=_artifact_dependency(artifacts["specs"]),
        design=_artifact_dependency(artifacts["design"]),
        tasks=_artifact_dependency(artifacts["tasks"]),
        apply=_apply_dependency(apply_state),
        verify=DEP_BLOCKED,
        archive=DEP_BLOCKED,
    )

    apply_progress_done = artifacts["applyProgress"] == ARTIFACT_DONE
    verify_report_done = artifacts["verifyReport"] == ARTIFACT_DONE

    if verify_report_done and core_ready and tasks.all_complete and verify_passing:
        deps.verify = DEP_ALL_DONE
    elif core_ready and (apply_state == APPLY_ALL_DONE or apply_progress_done):
        deps.verify = DEP_READY

    if deps.verify == DEP_ALL_DONE and tasks.all_complete:
        deps.archive = DEP_READY
    return deps


def _artifact_dependency(state: str) -> str:
    return DEP_ALL_DONE if state == ARTIFACT_DONE else DEP_BLOCKED


def _apply_dependency(apply_state: str) -> str:
    if apply_state == APPLY_READY:
        return DEP_READY
    if apply_state == APPLY_ALL_DONE:
        return DEP_ALL_DONE
    return DEP_BLOCKED


def _resolve_next_recommended(deps: Dependencies, apply_state: str) -> str:
    """Choose the single next action: apply when ready, else verify when ready,
    else archive when verify is done and apply is done, else resolve-blockers."""
    if deps.apply == DEP_READY:
        return PHASE_APPLY
    if deps.verify == DEP_READY:
        return PHASE_VERIFY
    if deps.verify == DEP_ALL_DONE and apply_state == APPLY_ALL_DONE:
        return PHASE_ARCHIVE
    return NEXT_RESOLVE_BLOCKERS


def _core_blocked_reasons(artifacts: dict[str, str], tasks: TaskProgress) -> list[str]:
    """List the human reasons the core is not ready: one per missing-or-partial
    core artifact, plus a note when tasks.md is present but has no checkboxes."""
    reasons: list[str] = []
    if artifacts["proposal"] != ARTIFACT_DONE:
        reasons.append("proposal.md is missing or partial.")
    if artifacts["specs"] != ARTIFACT_DONE:
        reasons.append("specs/**/spec.md is missing or partial.")
    if artifacts["design"] != ARTIFACT_DONE:
        reasons.append("design.md is missing or partial.")
    if artifacts["tasks"] != ARTIFACT_DONE:
        reasons.append("tasks.md is missing or partial.")
    if artifacts["tasks"] == ARTIFACT_DONE and tasks.total == 0:
        reasons.append("tasks.md has no markdown task checkboxes.")
    return reasons
